src/Lecture2.hs

```diff
@@ -48,7 +51,14 @@ zero, you can stop calculating product and return 0 immediately.
 84
 -}
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
+lazyProduct list = go list 1
+  where
+    go :: [Int] -> Int -> Int
+    go [] result = result
+    go (0 : _) _ = 0
+    go (x: xs) result = go xs x * result
```

Comment:
I have learned it in the next example
I use it because in Scala is very common to use the tail-recursion. It solves the StackOverFlow problem. I don't know how common it in Haskell

Reply:
The tail-recursion problem is not so big in Haskell because it's lazy and lots of functions tail-recursive automatically.
It's still possible to write non-tail-recursive implementation. For example, the following naive implementation of `product` is not tail-recursive:

```haskell
product [] = 1
product (x : xs) = x * product xs
```